opensrp-reveal/src/main/java/org/smartregister/reveal/presenter/ValidateUserLocationPresenter.java

```diff
@@ -0,0 +1,72 @@
+package org.smartregister.reveal.presenter;
+
+import android.location.Location;
+import android.os.Handler;
+import android.os.Looper;
+import android.util.Log;
+
+import com.mapbox.mapboxsdk.geometry.LatLng;
+
+import org.smartregister.reveal.BuildConfig;
+import org.smartregister.reveal.R;
+import org.smartregister.reveal.contract.UserLocationContract;
+import org.smartregister.reveal.contract.UserLocationContract.UserLocationCallback;
+import org.smartregister.reveal.contract.UserLocationContract.UserLocationView;
+
+/**
+ * Created by samuelgithengi on 2/13/19.
+ */
+public class ValidateUserLocationPresenter implements UserLocationContract.UserLocationPresenter {
+
+    private static final String TAG = "ValidateUserLocation";
+
+    private UserLocationView locationView;
+
+    private UserLocationCallback callback;
+
+    public ValidateUserLocationPresenter(UserLocationView locationView, UserLocationCallback callback) {
+        this.locationView = locationView;
+        this.callback = callback;
+    }
+
+    @Override
+    public void requestUserLocation() {
+        locationView.requestUserLocation();
+    }
+
+    @Override
+    public void onGetUserLocation(Location location) {
+        locationView.hideProgressDialog();
+        double offset = callback.getTargetCoordinates().distanceTo(
+                new LatLng(location.getLatitude(), location.getLongitude()));
+        if (offset > BuildConfig.MY_LOCATION_BUFFER) {
+            callback.requestUserPassword();
+        } else {
+            callback.onLocationValidated();
+        }
+    }
+
+    @Override
+    public void onGetUserLocationFailed() {
+        locationView.hideProgressDialog();
+        callback.requestUserPassword();
+    }
+
+    @Override
+    public void waitForUserLocation() {
+        Location location = locationView.getUserCurrentLocation();
+        Log.d(TAG, "user location: " + location);
+        if (location == null) {
+            locationView.showProgressDialog(R.string.narrowing_location_title, R.string.narrowing_location_message);
+            new Handler(Looper.getMainLooper()).postDelayed(new Runnable() {
+                @Override
+                public void run() {
+                    waitForUserLocation();
```

Comment:
Is it deterministic that at some point the location will not be null?

If not, this function could recurse indefinitely.

Reply:
Geowidget(Kujaku) registers for location updates whenever the location services are enabled. Google Fused Location client is used to provide the location updates.  Getting location may time especially if GPS has to be used but once location is retrieved the location will have the value.

Reply:
But this is on the UI thread, right?

So if that takes time then it may be an issue?

Reply:
I will put a timeout, where resolving location takes time it fails.

This does lock the UI thread since its putting a JOB on the UI thread looper that will be executed periodically. It does sleep the UI thread. The user is not allowed to interact with the app until resolving location times out or completes. 

Resolving the location is done on background threads on the fused client. What the UI thread just does is very light just checks if Geowidget has a location.

Reply:
I moved waiting to a different thread using a timer scheduler

Reply:
Cool.